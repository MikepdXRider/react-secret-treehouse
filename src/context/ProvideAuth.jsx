import { createContext, useState } from 'react';

export const authContext = createContext();

export function ProvideAuth({ children }) {
  const [user, setUser] = useState(null);

  const login = (email, password) => {
    const loginSuccessful =
      email === process.env.AUTH_EMAIL &&
      password === process.env.AUTH_PASSWORD;
    if (loginSuccessful) setUser({ email });
    return loginSuccessful;
  };

  const logout = (callback) => {
    setUser(null);
    callback();
  };

  return (
    <authContext.Provider value={{ user, login, logout }}>
      {children}
    </authContext.Provider>
  );
}
