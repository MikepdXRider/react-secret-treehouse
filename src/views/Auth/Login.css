.loginForm {
  max-width: 20vw;
  display: grid;
  grid: repeat(4, 1fr) / repeat(2, 1fr);
  gap: 1rem;
  align-items: center;
}

label {
  text-align: right;
}

button {
  grid-column: 1 / span 2;
  grid-row: 3 / span 2;
  height: 60%;
  border-radius: 5px;
  font-size: 1.1rem;
  cursor: pointer;
}

.error {
  color: rgb(253, 71, 85);
}
