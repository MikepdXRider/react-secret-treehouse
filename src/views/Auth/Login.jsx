import { useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useAuth } from '../../hooks/useAuth';
import { useForm } from '../../hooks/useForm';
import styles from './Login.css';

export default function Login() {
  const history = useHistory();
  const location = useLocation();
  const auth = useAuth();
  const { formState, handleFormChange } = useForm({ email: '', password: '' });
  const [error, setError] = useState(null);

  // The `from` property of `location.state` gives us
  // the URL to redirect to after logging in.
  const { from } = location.state || { from: { pathname: '/' } };

  const handleLogin = (event) => {
    event.preventDefault();
    const loginWasSuccessful = auth.login(formState.email, formState.password);

    // TODO: If login was unsuccessful, set an error with a message
    // to display to the user that their login failed.
    //
    // If login was successful, use the history hook
    // from React Router to replace the current URL with the URL
    // we need to redirect to.
    // See https://v5.reactrouter.com/web/api/history for the appropriate method to use
  };

  return (
    <>
      <h3>You must log in to view the page at {from.pathname}</h3>
      <form onSubmit={handleLogin} className={styles.loginForm}>
        <label>Email</label>
        <input
          id="email"
          name="email"
          type="email"
        />{' '}
        <label>Password</label>
        <input
          id="password"
          name="password"
          type="password"
        />
        <button type="submit" aria-label="Sign In">
          Sign in
        </button>
      </form>
      {error && <h4 className={styles.error}>{error}</h4>}
    </>
  );
}
