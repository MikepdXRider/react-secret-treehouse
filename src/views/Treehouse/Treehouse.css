.signOut {
  background-color: #ffce01;
  border: 1px solid #fcba12;
  color: #f24501;
  font-size: 1.2rem;
  padding: 7px 15px;
  text-transform: uppercase;
  font-weight: bold;
  margin: 1.5rem 0;
}
