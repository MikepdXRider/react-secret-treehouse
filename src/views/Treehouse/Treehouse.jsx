import { Link, useHistory } from 'react-router-dom';
import { useAuth } from '../../hooks/useAuth';
import secret from '../../assets/top-secret.png';
import styles from './Treehouse.css';

export default function Treehouse() {
  const history = useHistory();
  const { user, logout } = useAuth();

  const handleLogout = () => {
    logout(() => history.push('/'));
  };

  return (
    <>
      <h1>Welcome to the Secret Treehouse!</h1>
      <p>You are signed in as {user.email}</p>
      <Link to="/treehouse/members">View the Treehouse Members</Link>
      <button
        aria-label="Sign Out"
        type="button"
        className={styles.signOut}
        onClick={handleLogout}
      >
        Sign out
      </button>
      <img src={secret} alt="Top secret treehouse image" />
    </>
  );
}
