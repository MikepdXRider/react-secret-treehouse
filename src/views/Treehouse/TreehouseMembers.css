.members {
  max-width: 60vw;
}
