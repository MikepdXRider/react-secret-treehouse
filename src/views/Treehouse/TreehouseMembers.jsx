import { Link } from 'react-router-dom';
import styles from './TreehouseMembers.css';
import members from '../../assets/secret-treehouse-members.jpeg';

export default function TreehouseMembers() {
  return (
    <>
      <Link to="/treehouse">Go back to the Treehouse</Link>
      <img src={members} alt="Top secret treehouse members" className={styles.members} />
      <ul>
        <li>John</li>
        <li>Paul</li>
        <li>George</li>
        <li>Ringo</li>
      </ul>
    </>
  );
}
