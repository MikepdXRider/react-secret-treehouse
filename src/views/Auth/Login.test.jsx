import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { MemoryRouter, Route, Switch } from 'react-router-dom';
import { ProvideAuth } from '../../context/ProvideAuth';
import Login from './Login';
import Treehouse from '../Treehouse/Treehouse'

jest.mock('../../hooks/useAuth', () => {
  return {
    useAuth: function () {
      return {
        user: { email: 'test@example.com' },
        login: () => true,
      };
    },
  };
});

// Combined behavior & snapshot test
it('should allow the user to log in', async () => {
  const { container } = render(
    <ProvideAuth>
      <MemoryRouter initialEntries={["/login"]}>
        <Switch>
          <Route exact path="/login">
            <Login />
          </Route>
          <Route exact path="/">
            <Treehouse />
          </Route>
        </Switch>
      </MemoryRouter>
    </ProvideAuth>
  );

  expect(container).toMatchSnapshot();

  const emailField = screen.getByLabelText('Email');
  const passwordField = screen.getByLabelText('Password');
  const submitBtn = screen.getByRole('button', { name: 'Sign In' });
  
  fireEvent.change(emailField, {
    target: { value: 'test@example.com' },
  });
  fireEvent.change(passwordField, {
    target: { value: 'secret' },
  });
  
  expect(passwordField).toHaveValue('secret');
  expect(emailField).toHaveValue('test@example.com');

  fireEvent.click(submitBtn);

  return waitFor(() => {
    screen.getByText('Welcome to the Secret Treehouse!');
  });
});
