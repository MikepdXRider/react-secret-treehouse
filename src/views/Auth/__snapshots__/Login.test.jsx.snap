// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow the user to log in 1`] = `
<div>
  <h3>
    You must log in to view the page at 
    /
  </h3>
  <form
    class="loginForm"
  >
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      type="email"
      value=""
    />
     
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      type="password"
      value=""
    />
    <button
      aria-label="Sign In"
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
